host/xtest/aes_perf.c

```diff
@@ -0,0 +1,474 @@
+	fprintf(stderr, "  -m    AES mode: ECB, CBC, CTR, XTS [%s]\n",
+			mode_str(mode));
+	fprintf(stderr, "  -n    Outer loop iterations [%u]\n", n);
+	fprintf(stderr, "  -r    Get input data from /dev/urandom ");
+	fprintf(stderr, "(otherwise use zero-filled buffer)\n");
+	fprintf(stderr, "  -s    Buffer size (process <x> bytes at a time) ");
+	fprintf(stderr, "[%zu]\n", size);
+	fprintf(stderr, "  -v    Be verbose (use twice for greater effect)\n");
+	fprintf(stderr, "  -w    Warm-up time in seconds: execute a busy ");
+	fprintf(stderr, "loop before the test\n");
+	fprintf(stderr, "        to mitigate the effects of cpufreq etc. ");
+	fprintf(stderr, "[%u]\n", warmup);
+}
+
+static void alloc_shm(size_t sz, int in_place)
+{
+	TEEC_Result res;
+
+	in_shm.buffer = NULL;
+	in_shm.size = sz;
+	res = TEEC_AllocateSharedMemory(&ctx, &in_shm);
+	check_res(res, "TEEC_AllocateSharedMemory");
+
+	if (!in_place) {
+		out_shm.buffer = NULL;
+		out_shm.size = sz;
+		res = TEEC_AllocateSharedMemory(&ctx, &out_shm);
+		check_res(res, "TEEC_AllocateSharedMemory");
+	}
+}
+
+static void free_shm(void)
+{
+	TEEC_ReleaseSharedMemory(&in_shm);
+	TEEC_ReleaseSharedMemory(&out_shm);
+}
+
+static ssize_t read_random(void *in, size_t rsize)
+{
+	static int rnd;
+	ssize_t s;
+
+	if (!rnd) {
+		rnd = open("/dev/urandom", O_RDONLY);
+		if (rnd < 0) {
+			perror("open");
+			return 1;
+		}
+	}
+	s = read(rnd, in, rsize);
+	if (s < 0) {
+		perror("read");
+		return 1;
+	}
+	if ((size_t)s != rsize) {
+		printf("read: requested %zu bytes, got %zd\n",
+		       rsize, s);
+	}
+
+	return 0;
+}
+
+static void get_current_time(struct timespec *ts)
+{
+	if (clock_gettime(CLOCK_MONOTONIC, ts) < 0) {
+		perror("clock_gettime");
+		exit(1);
+	}
+}
+
+static uint64_t timespec_to_ns(struct timespec *ts)
+{
+	return ((uint64_t)ts->tv_sec * 1000000000) + ts->tv_nsec;
+}
+
+static uint64_t timespec_diff_ns(struct timespec *start, struct timespec *end)
+{
+	return timespec_to_ns(end) - timespec_to_ns(start);
+}
+
+static uint64_t run_test_once(void *in, size_t size, TEEC_Operation *op,
+			  int random_in)
+{
+	struct timespec t0, t1;
+	TEEC_Result res;
+	uint32_t ret_origin;
+
+	if (random_in)
+		read_random(in, size);
+	get_current_time(&t0);
+	res = TEEC_InvokeCommand(&sess, TA_AES_PERF_CMD_PROCESS, op,
+				 &ret_origin);
+	check_res(res, "TEEC_InvokeCommand");
+	get_current_time(&t1);
+
+	return timespec_diff_ns(&t0, &t1);
+}
+
+static void prepare_key(int decrypt, int keysize, int mode)
+{
+	TEEC_Result res;
+	uint32_t ret_origin;
+	TEEC_Operation op;
+
+	memset(&op, 0, sizeof(op));
+	op.paramTypes = TEEC_PARAM_TYPES(TEEC_VALUE_INPUT, TEEC_VALUE_INPUT,
+					 TEEC_NONE, TEEC_NONE);
+	op.params[0].value.a = decrypt;
+	op.params[0].value.b = keysize;
+	op.params[1].value.a = mode;
+	res = TEEC_InvokeCommand(&sess, TA_AES_PERF_CMD_PREPARE_KEY, &op,
+				 &ret_origin);
+	check_res(res, "TEEC_InvokeCommand");
+}
+
+static void do_warmup(int warmup)
```

Comment:
When I was first testing on HiKey I found this "warm-up" would reduce the value of "max" in the report (min/max/mean/stddev), and also slightly reduce stddev. The impact on the reported average time and throughput was not huge IIRC, which is expected given a sufficient number of loops (default value = 5000 seems more than enough to eliminate the effect of a few bogus values on start-up). But still, I was getting more stable results with warm-up enabled than without.'

Now I checked again and the effect is pretty much zero actually. Perhaps something changed in the kernel, like no frequency scaling? (I'm using our optee branch). Indeed, I don't see any `cpufreq` entry in the `/sys` filesystem:
```
root@HiKey:/ find /sys -name \*freq\*
root@HiKey:/
```